File: scripts/dev/generate_characterization_tests.py
#!/usr/bin/env python3
import argparse
import os
import re
import sys

def parse_header(header_path):
    with open(header_path, 'r') as f:
        content = f.read()

    # Regex for finding interfaces
    # Capture the class name
    interface_pattern = re.compile(r'@interface\s+(\w+)\s*:')
    
    classes = {}
    
    # helper to parse methods inside an @interface ... @end block?
    # Actually, simpler to just scan the whole file for @implementation or @interface and methods.
    # But headers usually validly separate them.
    
    # Let's iterate over the file and track which class we are in.
    # But often interfaces are defined sequentially.
    
    # Simplified approach: Extract blocks for each interface
    # This is hard with regex.
    # Failure mode: just find all methods and try to guess? No.
    
    # Better approach:
    # 1. Find all `@interface Name : Super`
    # 2. Find the END of that block (next @interface or @citation or @end)
    # 3. Parse methods within that chunk.
    
    # Let's try to match `@interface ... @end` blocks loosely.
    
    # split by @interface
    chunks = re.split(r'(@interface\s+\w+)', content)
    
    current_class = None
    
    for chunk in chunks:
        # Check if this chunk is the start of a class definition
        class_match = re.match(r'@interface\s+(\w+)', chunk)
        if class_match:
             # This chunk IS the marker, the NEXT chunk is the content
             current_class = class_match.group(1)
             continue
             
        if current_class:
            # This chunk contains the body of current_class
            # Parse methods
            
            # Instance methods (-) and Class methods (+)
            method_pattern = re.compile(r'^\s*([+-])\s*\(([^)]+)\)([^;]+);', re.MULTILINE)
            
            cls_methods = []
            for match in method_pattern.finditer(chunk):
                kind = match.group(1) # + or -
                return_type = match.group(2).strip()
                signature_raw = match.group(3).strip()
                
                parts = signature_raw.split(':')
                first_part = parts[0].strip()
                
                cls_methods.append({
                    'kind': kind,
                    'return_type': return_type,
                    'full_text': match.group(0).strip(),
                    'name': first_part
                })
            
            classes[current_class] = cls_methods
            current_class = None # Reset
            
    return classes

def generate_test_content(class_name, methods, header_path):
    # Determine looking relative path for import
    # header_path is absolute. We need to find where it is relative to Sources/
    
    import_path = header_path
    if "Garazyk/Sources/" in header_path:
        import_path = header_path.split("Garazyk/Sources/")[1]
    
    content = f"""#import "CharacterizationTestBase.h"
#import "{import_path}"

@interface {class_name}CharacterizationTests : CharacterizationTestBase

@property (nonatomic, strong) {class_name} *subject;

@end

@implementation {class_name}CharacterizationTests

- (void)setUp {{
    [super setUp];
    // TODO: Initialize self.subject
    // self.subject = [[{class_name} alloc] init];
}}

- (void)tearDown {{
    self.subject = nil;
    [super tearDown];
}}

/*
 * Characterization Tests for {class_name}
 * Generated automatically. Please implement specific scenarios.
 */
"""

    seen_methods = set()
    for method in methods:
        method_name = method['name']
        kind_prefix = "Class_" if method['kind'] == '+' else ""
        test_method_name = f"testCharacterization_{kind_prefix}{method_name}"
        
        # Handle overloads/duplicates in naming
        counter = 1
        base_name = test_method_name
        while test_method_name in seen_methods:
            counter += 1
            test_method_name = f"{base_name}_{counter}"
        seen_methods.add(test_method_name)
        
        act_line = ""
        if method['kind'] == '+':
            act_line = f"// [{class_name} {method_name}...];"
        else:
            act_line = f"// [self.subject {method_name}...];"

        content += f"""
- (void){test_method_name} {{
    /* Target Method:
     {method['full_text']}
    */
    
    // 1. Arrange
    
    // 2. Act
    {act_line}
    
    // 3. Assert
    // XCTFail(@"Test not implemented");
}}
"""

    content += "\n@end\n"
    return content

def main():
    parser = argparse.ArgumentParser(description='Generate characterization tests for a class.')
    parser.add_argument('header', help='Path to the header file of the class')
    parser.add_argument('--output', help='Output directory for the test file', default='Garazyk/Tests/CharacterizationTests')
    parser.add_argument('--target-class', help='Specific class to generate tests for', default=None)
    
    args = parser.parse_args()
    
    if not os.path.exists(args.header):
        print(f"Error: Header file not found: {args.header}")
        sys.exit(1)
        
    classes = parse_header(args.header)
    
    if not classes:
        print("Error: Could not find any @interface definitions in header.")
        sys.exit(1)
        
    # Determine target class
    target = args.target_class
    if not target:
        # derive from filename
        basename = os.path.basename(args.header)
        root = os.path.splitext(basename)[0]
        if root in classes:
            target = root
        else:
            # Pick the first one if exact match not found?
            # Or just warn?
            # Let's pick the one with the most methods? 
            # Or just the first one found.
            target = list(classes.keys())[0]
            print(f"Warning: Class '{root}' not found in header. Defaulting to '{target}'.")
            
    if target not in classes:
        print(f"Error: Target class '{target}' not found in {list(classes.keys())}")
        sys.exit(1)
        
    methods = classes[target]
    print(f"Generating tests for class: {target} with {len(methods)} methods.")
    
    test_content = generate_test_content(target, methods, args.header)
    
    output_filename = f"{target}CharacterizationTests.m"
    output_path = os.path.join(args.output, output_filename)
    
    # Create directory if it doesn't exist
    os.makedirs(args.output, exist_ok=True)
    
    if os.path.exists(output_path):
        print(f"Warning: {output_path} already exists. Skipping generation to avoid overwrite.")
    else:
        with open(output_path, 'w') as f:
            f.write(test_content)
        print(f"Generated {output_path}")

if __name__ == '__main__':
    main()

File: scripts/dev/generate_characterization_tests.ts
#!/usr/bin/env -S deno run -A
import { basename, dirname, join, relative, resolve } from "@std/path";

interface MethodInfo {
  kind: "+" | "-";
  returnType: string;
  fullText: string;
  name: string;
}

interface Args {
  header: string;
  output: string;
  targetClass?: string;
}

function parseArgs(argv: string[]): Args {
  let header = "";
  let output = "Garazyk/Tests/CharacterizationTests";
  let targetClass: string | undefined;
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--output") output = argv[++i] ?? output;
    else if (arg === "--target-class") targetClass = argv[++i];
    else if (!header) header = arg;
    else {
      console.error(`Unexpected argument: ${arg}`);
      Deno.exit(2);
    }
  }
  if (!header) {
    console.error(
      "Usage: scripts/dev/generate_characterization_tests.ts HEADER [--output DIR] [--target-class CLASS]",
    );
    Deno.exit(2);
  }
  return { header, output, targetClass };
}

async function parseHeader(headerPath: string): Promise<Record<string, MethodInfo[]>> {
  const content = await Deno.readTextFile(headerPath);
  const chunks = content.split(/(@interface\s+\w+)/);
  const classes: Record<string, MethodInfo[]> = {};
  let currentClass: string | undefined;

  for (const chunk of chunks) {
    const classMatch = chunk.match(/^@interface\s+(\w+)/);
    if (classMatch) {
      currentClass = classMatch[1];
      continue;
    }
    if (!currentClass) continue;

    const methods: MethodInfo[] = [];
    const methodPattern = /^\s*([+-])\s*\(([^)]+)\)([^;]+);/gm;
    for (const match of chunk.matchAll(methodPattern)) {
      const signature = match[3].trim();
      methods.push({
        kind: match[1] as "+" | "-",
        returnType: match[2].trim(),
        fullText: match[0].trim(),
        name: signature.split(":")[0].trim(),
      });
    }
    classes[currentClass] = methods;
    currentClass = undefined;
  }

  return classes;
}

function importPathForHeader(headerPath: string): string {
  const normalized = resolve(headerPath);
  const marker = "Garazyk/Sources/";
  const idx = normalized.indexOf(marker);
  return idx >= 0 ? normalized.slice(idx + marker.length) : basename(headerPath);
}

function generateTestContent(className: string, methods: MethodInfo[], headerPath: string): string {
  const importPath = importPathForHeader(headerPath);
  let content = `#import "CharacterizationTestBase.h"
#import "${importPath}"

@interface ${className}CharacterizationTests : CharacterizationTestBase

@property (nonatomic, strong) ${className} *subject;

@end

@implementation ${className}CharacterizationTests

- (void)setUp {
    [super setUp];
    // TODO: Initialize self.subject
    // self.subject = [[${className} alloc] init];
}

- (void)tearDown {
    self.subject = nil;
    [super tearDown];
}

/*
 * Characterization Tests for ${className}
 * Generated automatically. Please implement specific scenarios.
 */
`;

  const seen = new Set<string>();
  for (const method of methods) {
    const kindPrefix = method.kind === "+" ? "Class_" : "";
    const baseName = `testCharacterization_${kindPrefix}${method.name}`;
    let testName = baseName;
    let counter = 1;
    while (seen.has(testName)) testName = `${baseName}_${++counter}`;
    seen.add(testName);
    const actLine = method.kind === "+"
      ? `// [${className} ${method.name}...];`
      : `// [self.subject ${method.name}...];`;

    content += `
- (void)${testName} {
    /* Target Method:
     ${method.fullText}
    */

    // 1. Arrange

    // 2. Act
    ${actLine}

    // 3. Assert
    // XCTFail(@"Test not implemented");
}
`;
  }

  return `${content}\n@end\n`;
}

async function main() {
  const args = parseArgs(Deno.args);
  try {
    await Deno.stat(args.header);
  } catch {
    console.error(`Error: Header file not found: ${args.header}`);
    Deno.exit(1);
  }

  const classes = await parseHeader(args.header);
  const names = Object.keys(classes);
  if (names.length === 0) {
    console.error("Error: Could not find any @interface definitions in header.");
    Deno.exit(1);
  }

  let target = args.targetClass;
  if (!target) {
    const root = basename(args.header).replace(/\.[^.]+$/, "");
    target = classes[root] ? root : names[0];
    if (target !== root) {
      console.log(`Warning: Class '${root}' not found in header. Defaulting to '${target}'.`);
    }
  }
  if (!classes[target]) {
    console.error(`Error: Target class '${target}' not found in ${names.join(", ")}`);
    Deno.exit(1);
  }

  console.log(`Generating tests for class: ${target} with ${classes[target].length} methods.`);
  const outputPath = join(args.output, `${target}CharacterizationTests.m`);
  await Deno.mkdir(dirname(outputPath), { recursive: true });
  try {
    await Deno.stat(outputPath);
    console.log(`Warning: ${outputPath} already exists. Skipping generation to avoid overwrite.`);
    return;
  } catch {
    // Good: the output does not exist.
  }
  await Deno.writeTextFile(outputPath, generateTestContent(target, classes[target], args.header));
  console.log(`Generated ${relative(Deno.cwd(), outputPath)}`);
}

if (import.meta.main) {
  await main();
}
